### DISCUSSION TITLE
[QUESTION] callback_data from http api /sendMessage not triggered

### DISCUSSION THREAD

**Discussion by vbeskrovny** [python-telegram-bot/python-telegram-bot#3400] [Feature Suggestions]
### Issue I am facing

Hi all, sorry, I do believe my question is very simple or stupid, but I cannot overcome it myself.

I am sending a question to the user's chat with the inline markup keyboard, having two buttons "YES" & "NO" with the same callback_data values 'YES' and 'NO'.

I do receive in the bot on my phone the question with the keyboard, however, if I click the button nothing happens. I see the rotating circle in the upper right corner of the button and this is it.

When the question is triggered via /some_command scripted inside the bot - all works.

What could be wrong? Any ideas?

### Operating System

Debian

### Version of Python, python-telegram-bot & dependencies

```shell
Python3.7
```

**Comment by Bibo-Joshi** [comment]
@vbeskrovny Hi. When you press an inline button, Telegram sends an update with <a href="https://docs.python-telegram-bot.org/telegram.update.html#telegram.Update.callback_query">telegram.Update.callback_query</a> <a href="https://core.telegram.org/bots/api#update">ᵀᴱᴸᴱᴳᴿᴬᴹ</a> to your bot. You will have to handle that update. See <a href="https://docs.python-telegram-bot.org/examples.html#examples-inlinekeyboard">inlinekeyboard.py</a> for a simple example.

**Comment by vbeskrovny** [comment]
Thanks for the prompt response.
Sure, I do have two handlers:

                CallbackQueryHandler(q_accepted, pattern="^YES$"),
                CallbackQueryHandler(q_ignored, pattern="^NO$"),

Or do you think I have to create a "generic" handler without filters?

**Reply by Bibo-Joshi** [reply]
That's impossible to tell from the available info … Please provide a complete [MWE](../wiki/MWE).

**Reply by vbeskrovny** [reply]
Here is my bot app code:
```python
#!/usr/bin/env python3

import os
import sys
import time
import inspect
import requests
import logging
import json
from pprint import pprint


# # from apscheduler.schedulers.background import BackgroundScheduler

import asyncio
import html
from dataclasses import dataclass
from http import HTTPStatus

import uvicorn
from starlette.applications import Starlette
from starlette.requests import Request
from starlette.responses import PlainTextResponse, Response
from starlette.routing import Route



from telegram.constants import ParseMode
from telegram import InlineKeyboardButton, InlineKeyboardMarkup, Update
from telegram.ext import (
	Application,
	CallbackQueryHandler,
	CommandHandler,
	ContextTypes,
	ConversationHandler,
	TypeHandler,
	ExtBot,
	CallbackContext,
)






SELF = os.path.realpath(sys.argv[0])
TOKEN = 'TOKEN_GOES_HERE'


logging.basicConfig(format="%(asctime)s - %(name)s - %(levelname)s - %(message)s", level=logging.INFO)
logger = logging.getLogger(__name__)


def log(txt):
	logger.info(txt)



# Stages
START_ROUTES, END_ROUTES = range(2)




async def start(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
	log('Start...')




async def generic_handler(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
	log('Generic...')
	query = update.callback_query
	await query.answer()
	await query.edit_message_text(text=f"Selected option: {query.data}")





def main() -> None:
	
    application = Application.builder().token(TOKEN).build()
    conv_handler = ConversationHandler(
        entry_points=[CommandHandler("start", start)],
        states={
            START_ROUTES: [
                CallbackQueryHandler(generic_handler),
            ],
        },
        fallbacks=[CommandHandler("start", start)],
    )
    # Add ConversationHandler to application that will be used for handling updates
    application.add_handler(conv_handler)
    
    # Run the bot until the user presses Ctrl-C
    application.run_polling()


if __name__ == "__main__":
	main()

	
	
```

And here is  curl request:

```
<?php

				$bot_token = 'TOKEN_GOES_HERE';
				$chat_id = 'here_is_my_user_id';

				$keyboard = array([
					[ 'text' => ' YES', 'callback_data'=> 'YES' ],
					[ 'text' => 'NO', 'callback_data'=> 'NO' ],
				]);
    
				$inlineKeyboardMarkup = array('inline_keyboard' => $keyboard);


				$url = 'https://api.telegram.org/bot'.$bot_token.'/sendMessage';
				
				$ch = curl_init();
				curl_setopt($ch, CURLOPT_URL, $url);
				curl_setopt($ch, CURLOPT_POST, 1);
				curl_setopt($ch, CURLOPT_POSTFIELDS, http_build_query([
						'chat_id' => $chat_id,
						'text' => 'YES/NO ?',
						'reply_markup' => json_encode($inlineKeyboardMarkup)
					]));
				curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
				$result = json_decode(curl_exec($ch), true);
				curl_close ($ch);

```
PS: I appreciate your involvement...

**Reply by Bibo-Joshi** [reply]
The `CallbackQueryHandler` is in the state `START_ROUTES` of a `ConversationHandler`. In only ever has the chance to handle an update if the conversation is in that state. As your example is incomplete, I can only guess that the `start` callback is supposed to `return START_ROUTES`, but if that has not happened, the `CallbackQueryHandler` is not even asked to check if it wants to handle the update.

If you need more help, please provide an actually complete example that I can run completely in python.

**Reply by vbeskrovny** [reply]
Well... That's exactly the complete example I have. The start callback is not used and not important. What I cannot trigger generic_handler from markup keyboard sent from curl... That is the issue I don't understand how to resolve...

**Reply by vbeskrovny** [reply]
The task I am trying to solve is fairly simple, I suppose. To get a choice of two buttons with the callbacks from curl request and let the bot handle these callbacks...

**Reply by vbeskrovny** [reply]
Here is what I try to reproduce: https://stackoverflow.com/questions/54073084/telegram-bot-on-python-with-callback
But the callback_data is not handled...

**Reply by Bibo-Joshi** [reply]
> The start callback is not used and not important.

Then you're misunderstanding `ConversationHandler`. Please make sure to read the documenation of that handler as well as to check out the examples that we have on that hanlder.

**Reply by vbeskrovny** [reply]
> > The start callback is not used and not important.
> 
> Then you're misunderstanding `ConversationHandler`. Please make sure to read the documenation of that handler as well as to check out the examples that we have on that hanlder.

Do you mean, that the callback_data is not handled because of "start" callback?

**Reply by Bibo-Joshi** [reply]
Yes, that's what I was elaborating in https://github.com/python-telegram-bot/python-telegram-bot/discussions/3400#discussioncomment-4264083

**Comment by vbeskrovny** [comment]
Alright, will try to find a workaround somehow. Thanks!

**Reply by Bibo-Joshi** [reply]
I suggest removing the `ConversationHandler` and simply adding the `CallbackQueryHandler` directly to the `application`. If that does not give the behavior that you'd like to have, feel free to comment again.

**Reply by vbeskrovny** [reply]
Awesome :)
